### PR TITLE
Add pre-request tab and minimal execution context

### PR DESCRIPTION
**Pull request by jackkav**
inspired by: https://www.electronjs.org/docs/latest/tutorial/message-ports/#worker-process

And #6991

provides a promise based API for the insomnia renderer to send a script and request related data to a hidden browser window and receive the request data back mutated by the script.

Goals:
- to use a similar sandboxing mechanism to insomnia renderer, where we remove require from the render and use preload to allow desired node modules.
- use a single channel promise to perform operations.
- add a basic ux and e2e test
- low code mechanics to make it easier to change

todo

- [x] run code on another browser and return a promise
- [x] wire up script tab to request
- [x] transform request using promise
- [x] e2e test adding a header to a request with a script
- [x] errors in the script should be visible in response pane
- [x] console logs should appear in the timeline, either write from worker or return logs with context
  - delaying until next PR because it requires a network rewrite to make timelines append-able for http requests, eg deterministic names and newlines separated json
- [x] can require allow listed modules
- [x] add node built in crypto
- [x] reload hidden window on change -> remove preload / modify window-utils
- [x] investigate context isolation

future work
- [ ] timeout and cancel and close()?
- [ ] clean up listeners, requires a close event in the hidden window which might not be needed?
- [ ] find instability triggers, test and handle them

closes INS-3380

### CODE REVIEW

File: packages/insomnia/src/main/window-utils.ts [hunk 1]
```diff
@@ -68,14 +69,13 @@ export async function createHiddenBrowserWindow(): Promise<ElectronBrowserWindow
...
   }
   const hiddenBrowserWindow = new BrowserWindow({
-    show: false,
+    show: process.env.NODE_ENV === 'development',
     title: 'HiddenBrowserWindow',
     webPreferences: {
-      sandbox: true,
-      contextIsolation: true,
-      nodeIntegration: false,
+      contextIsolation: false,
```
Comment (ihexxa) [inline packages/insomnia/src/main/window-utils.ts]: This is awesome to simplify parts such as message channel creation and message handling in preload. And I would propose to enable those flags if possible, as this environment is mainly for those anonymous code and you know it would be difficult for us to turn on them again :). I also spent lots of time to keep them enabled in another PR, please consider this.

File: packages/insomnia/src/main/window-utils.ts [hunk 2]
```diff
@@ -93,7 +93,14 @@ export async function createHiddenBrowserWindow(): Promise<ElectronBrowserWindow
       browserWindows.delete('HiddenBrowserWindow');
     }
   });
+  const mainWindow = browserWindows.get('Insomnia');
 
+  invariant(mainWindow, 'mainWindow is not defined');
+  mainWindow.webContents.mainFrame.ipc.on('open-channel-to-hidden-browser-window', event => {
+    const { port1, port2 } = new MessageChannelMain();
```
Comment (ihexxa) [inline packages/insomnia/src/main/window-utils.ts:99]: We may `close` these ports to avoid leaking and take care of crash of the window, as they are related to stability and user experience.

File: packages/insomnia/src/main/window-utils.ts (hunk 1 above)
Comment (jackkav) [reply packages/insomnia/src/main/window-utils.ts]: This is an idea for a developer experience tradeoff. I have also spent many hours writing bridges in the app to prepare for contextIsolation. Its something I want to investigate but it requires more boilerplate and I don't want to add anymore until I can deliver a more seamless dx. The faster it is to work on the more secure we can make it.
So far I enabled hot reload of changes to the hidden window index.ts on change without restarting the app. The next step is to stop vite from reloading the main browser window when we change the hidden window.
After this is will be easier to experiment with the security flags

Comment (jackkav) [reply packages/insomnia/src/main/window-utils.ts]: Ok thanks for the push @ihexxa ive reenabled contextIsolation, and pointed the index.html directly at the index.ts. The dx is improved and only preload changes require a full reload. However i still havent stopped insomnia from reloading on hidden window changes but i found a workaround. If you logout it doesnt go back to workspace view at reload so the learning loop is faster.

File: packages/insomnia/src/main/window-utils.ts (hunk 2 above)
Comment (jackkav) [reply packages/insomnia/src/main/window-utils.ts:99]: I want to scope these meta aspects out into testable PRs where possible ill add it to future work

**Review by gatzjames — DISMISSED** [review]

**Review by gatzjames — APPROVED** [review]

**Review by filfreire — APPROVED** [review]

File: packages/insomnia/src/main/window-utils.ts (hunk 2 above)
Comment (ihexxa) [reply packages/insomnia/src/main/window-utils.ts:99]: :+1: thanks for understanding, it also makes sense to me and let's test it later as probably it is not necessary.
